server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```diff
@@ -82,51 +93,128 @@ public void onNext(final StreamRequest request) {
-                        .onError(responseObserver)
+                        .onErrorHandler(this::onError)
+                        .require(requirements)
                         .submit(() -> {
-                            Object[] objs = referenceObjects.stream().map(ExportObject::get).toArray();
-                            // TODO: Should we try/catch this to recover/not close from plugin errors? Or make the
-                            // client
-                            // recover? Would be useful if the plugin is complex
-                            messageStream.onData(data.getPayload().asReadOnlyByteBuffer(), objs);
+                            // Run the specified work. Note that we're not concerned about exceptions, the stream will
+                            // be dead (via onError) and won't be used again.
+                            try {
+                                operation.run();
+                            } catch (ObjectType.ObjectCommunicationException e) {
+                                throw new IllegalStateException("Error occurred communicating with client", e);
+                            }
+
+                            // Set running to false (it must be true at this time) so that any new work can race being
+                            // added
+                            running.set(false);
+                            doWork();
                         });
+            };
+            // gRPC guarantees we can't race in this method
+            operations.add(runnable);
+            if (running.compareAndSet(false, true)) {
+                // no other thread is actually working, we can race to start work on this thread
+                doWork();
             }
         }
 
+        private void doWork() {
+            // More than one thread can arrive here at the same time, but only one will pass the compareAndSet
+            StreamOperation next = operations.peek();
+
+            // If we fail the null check, no work to do, leave running false (though if work was added right after
+            // peek(), that thread will make it into here and start). If we fail the running check, something else has
+            // already started work
+            if (next != null && running.compareAndSet(false, true)) {
+                // We have the running lock again, and should remove the item we just peeked at
+                StreamOperation actualNext = operations.poll();
+                Assert.eq(next, "next", actualNext, "actualNext");
+
+                // Run the new item
+                try {
+                    next.run();
+                } catch (ObjectType.ObjectCommunicationException e) {
+                    // Can only be called from within a context where exceptions will end the stream, so wrap and throw
+                    throw new IllegalStateException("Error occurred communicating with client", e);
+                }
+            }
+        }
+
+        private void fail() {
```

Comment:
unused.

Reply:
yep, was just notes for me to follow up on, will be gone in next commit